### ISSUE TITLE
reportOptionalMemberAccess when check is in subcall and determines the code path

### ISSUE THREAD

**Issue by thfr** [microsoft/pyright#3191]
**Description**
The optional member access check seems to be only applied if the check happens in the same scope. If the check happens outside of the current scope,  `reportOptionalMemberAccess` is issued by Pyright. I belief this is due to the dynamic nature of Python methods but I have not a found an issue about this topic directly. The only one that came close was #2040.

**Example code**
```python
import random
import typing

optional_str: typing.Optional[str] = None

if random.random() > 0.5:
    optional_str = "This is your chance"


def check_optional_str() -> bool:
    return optional_str is not None


def check_optional_str_arg(optional_str: typing.Optional[str]) -> bool:
    return optional_str is not None


if check_optional_str():
    print(optional_str.upper())

if check_optional_str_arg(optional_str):
    print(optional_str.upper())

if optional_str is not None:
    print(optional_str.upper())
```

**Pyright Output**

```
No configuration file found.
stubPath /tmp/typings is not a valid directory.
No pyproject.toml file found.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/tmp/pyright_optional_check.py
  /tmp/pyright_optional_check.py:19:24 - error: "upper" is not a known member of "None" (reportOptionalMemberAccess)
  /tmp/pyright_optional_check.py:22:24 - error: "upper" is not a known member of "None" (reportOptionalMemberAccess)
2 errors, 0 warnings, 0 informations 
Completed in 0.654sec
```

**Expected behavior**

The report should not be issued if the check occurs in function call and determines the code path.

**VS Code extension or command-line**

```
$ pyright --version
pyright 1.1.229
```

**Additional context**

Possibly related issue #2040

Thanks for developing Pyright and sorry if this turns out to be by design.

**Comment by erictraut** [comment]
Yes, the current behavior is correct and intended.

[Type narrowing](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-narrowing) is not applied across functions unless you use a user-defined type guard function as documented in [PEP 647](https://peps.python.org/pep-0647/).

To build upon your example, you could define `check_optional_str_arg` as follows:

```python
def check_optional_str_arg(optional_str: typing.Optional[str]) -> typing.TypeGuard[str]:
    return optional_str is not None
```

**Comment by thfr** [comment]
@erictraut Thanks for short explanation and the helpful links. I think, I am starting to wrap my head around this concept. Again, sorry for creating this issue.

One important thing that I took from this and was the following code that solves the annoyances that I was struggling with:
```python
class CheckOptionalStr:
    def __init__(self, string: typing.Optional[str]) -> None:
        self._optional_str: typing.Optional[str] = string
        self._rand_num = random.random()

    # does type narrowing and checks another condition
    def check(self) -> typing.TypeGuard[str]:
        return self._optional_str is not None and self._rand_num > 0.5

    def print(self):
        if self.check():
            print(self._optional_str)
        else:
            print("Nothing to see here")
```

**Comment by thfr** [comment]
I do not get reports from Pyright about it, is that a bug? 

Here is the full example and Pyright's output:
```python
import random
import typing

optional_str: typing.Optional[str] = None

if random.random() > 0.5:
    optional_str = "This is your chance"


class CheckOptionalStr:
    def __init__(self, string: typing.Optional[str]) -> None:
        self._optional_str: typing.Optional[str] = string
        self._rand_num = random.random()

    # does type narrowing and checks another condition
    def check(self) -> typing.TypeGuard[str]:
        return self._optional_str is not None and self._rand_num > 0.5

    def print(self):
        if self.check():
            print(self._optional_str)
        else:
            print("Nothing to see here")


cos = CheckOptionalStr(optional_str)
if cos.check():
    print(cos._optional_str)
```
```
No configuration file found.
stubPath /tmp/typings is not a valid directory.
No pyproject.toml file found.
Assuming Python platform Linux
Searching for source files
Found 1 source file
0 errors, 0 warnings, 0 informations 
Completed in 0.616sec
```
```
$ python3 --version
Python 3.10.2
$ pyright --version
pyright 1.1.229
```
